drgn_tools/multipath.py

```diff
@@ -0,0 +1,330 @@
+# Copyright (c) 2024, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helpers for multipath
+"""
+import argparse
+from typing import Iterable
+
+from drgn import cast
+from drgn import Object
+from drgn import Program
+from drgn import Type
+from drgn.helpers.linux.list import list_for_each_entry
+
+from drgn_tools.block import blkdev_name
+from drgn_tools.block import blkdev_ro
+from drgn_tools.block import blkdev_size
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.dm import dm_table
+from drgn_tools.dm import dm_target_name
+from drgn_tools.dm import for_each_dm
+from drgn_tools.module import ensure_debuginfo
+from drgn_tools.util import BitNumberFlags
+from drgn_tools.util import has_member
+from drgn_tools.util import type_lookup_conflict
+
+dm_device_count = 1
+
+
+class DmMultipathStateFlags(BitNumberFlags):
+    MPATHF_QUEUE_IO = 0
+    MPATHF_QUEUE_IF_NO_PATH = 1
+    MPATHF_SAVED_QUEUE_IF_NO_PATH = 2
+    MPATHF_RETAIN_ATTACHED_HW_HANDLER = 3
+    MPATHF_PG_INIT_DISABLED = 4
+    MPATHF_PG_INIT_REQUIRED = 5
+    MPATHF_PG_INIT_DELAY_RETRY = 6
+
+
+def for_each_priority_group(multipath: Object) -> Iterable[Object]:
+    """
+    Iterates through priority groups of a multipath device.
+
+    Paths are grouped into priority groups.
+    A multipath device can have multiple priority groups.
+
+    :param multipath: `struct multipath *`
+
+    :returns: Iterable of `strut priority_group *`
+    """
+    return list_for_each_entry(
+        "struct priority_group",
+        multipath.priority_groups.address_of_(),
+        "list",
+    )
+
+
+def for_each_pgpath(priority_group: Object) -> Iterable[Object]:
+    """
+    Iterates through pgpaths for each priority_group.
+
+    struct pgpath contain path properties.
+
+    :param priority_group: `struct priority_group *`
+
+    :returns: Iterable of `struct pgpath *`
+    """
+    return list_for_each_entry(
+        "struct pgpath", priority_group.pgpaths.address_of_(), "list"
+    )
+
+
+def switch_header(ps_type: str) -> None:
+    """
+    Format the output header as per ps policy.
+
+    :param  ps_type: Path selection policy name.
+
+    :return None.
+    """
+    header_mapping = {
+        "service-time": "   InFlightIO_Size   RThroughput  RCount  Load",
+        "round-robin": "   Repeat_Count",
+        "queue-length": "   Repeat_Count  Queue_Length",
+        "io-affinity": "   Refcount   failed",
+        "historical-service-time": "Feature not supported for "
+        "historical-service-time.",
```

Comment:
"historical-service-time" is not supported, we don't need it here.

Reply:
Most of the code for hystorical-serivce-time is a place holder for this policy for future. 
I added check for path selector at the very beginning of code is just to avoid unnecessary  execution of code.

Reply:
Ok, if plan to support it in future, then keep it.